omnigibson/action_primitives/starter_semantic_action_primitives.py

```diff
@@ -0,0 +1,1898 @@
+            yield self.with_context(action)
+
+            body_target_pose = self._get_pose_in_robot_frame(end_pose)
+            diff_yaw = T.wrap_angle(T.quat2euler(body_target_pose[1])[2])
+        
+        empty_action = self._empty_action(arm_pose_to_keep=initial_eef_pos)
+        yield self.with_context(empty_action)
+            
+    def _sample_pose_near_object(self, obj, pose_on_obj=None, **kwargs):
+        """
+        Returns a 2d pose for the robot within in the range of the object and where the robot is not in collision with anything
+
+        Args:
+            obj (StatefulObject): Object to sample a 2d pose near
+            pose_on_obj (Iterable of arrays or None): The pose to sample near
+
+        Returns:
+            2-tuple:
+                - 3-array: (x,y,z) Position in the world frame
+                - 4-array: (x,y,z,w) Quaternion orientation in the world frame
+        """
+        with UndoableContext(self.robot, self.robot_copy, "simplified") as context:
+            for _ in range(MAX_ATTEMPTS_FOR_SAMPLING_POSE_NEAR_OBJECT):
+                if pose_on_obj is None:
+                    pos_on_obj = self._sample_position_on_aabb_face(obj)
+                    pose_on_obj = [pos_on_obj, np.array([0, 0, 0, 1])]
+
+                distance = np.random.uniform(0.0, 5.0)
+                yaw = np.random.uniform(-np.pi, np.pi)
+                avg_arm_workspace_range = np.mean(self.robot.arm_workspace_range[self.arm])
+                pose_2d = np.array(
+                    [pose_on_obj[0][0] + distance * np.cos(yaw), pose_on_obj[0][1] + distance * np.sin(yaw), yaw + np.pi - avg_arm_workspace_range]
+                )
+                # Check room
+                obj_rooms = obj.in_rooms if obj.in_rooms else [self.scene._seg_map.get_room_instance_by_point(pose_on_obj[0][:2])]
+                if self.scene._seg_map.get_room_instance_by_point(pose_2d[:2]) not in obj_rooms:
+                    indented_print("Candidate position is in the wrong room.")
+                    continue
+
+                if not self._test_pose(pose_2d, context, pose_on_obj=pose_on_obj, **kwargs):
+                    continue
+
+                return pose_2d
+
+            raise ActionPrimitiveError(
+                ActionPrimitiveError.Reason.SAMPLING_ERROR, "Could not find valid position near object.",
+                {"target object": obj.name, "target pos": obj.get_position(), "pose on target": pose_on_obj}
+            )
+
+    @staticmethod
+    def _sample_position_on_aabb_face(target_obj):
+        """
+        Returns a position on the axis-aligned bounding box (AABB) faces of the target object.
+
+        Args:
+            target_obj (StatefulObject): Object to sample a position on
+
+        Returns:
+            3-array: (x,y,z) Position in the world frame
+        """
+        aabb_center, aabb_extent = target_obj.aabb_center, target_obj.aabb_extent
+        # We want to sample only from the side-facing faces.
```

Comment:
why only the side? is this for the open / close primitive?

Reply:
It's for navigating - generally want to pick an approach side. Reflected in comments